File: harness/py-0.8-httpx/smoke_test.py
"""Local smoke for this httpx client entry: it builds the go-0.12 reference
server from the sibling entry (the known-good verifier, like the orchestrator
pairs entries), starts it on a loopback port against the committed fixture
(one process per mode, so each mode gets a fresh replay cache), and drives
every HTTP-applicable scenario from the repository's scenarios.yaml through
this entry's client runnable, judging its printed outcome line. This proves
the entry without containers; the Dockerfile still owns the pinned
interpreter.

Run it inside the entry's venv — smoke.sh builds one with the exact pins and
invokes this file. The entry id (and so the client runnable's name) is the
entry directory's own name.
"""

from __future__ import annotations

import json
import signal
import subprocess
import sys
import tempfile
import threading
from collections.abc import Iterator
from contextlib import contextmanager
from pathlib import Path
from typing import Any, NoReturn

import yaml

ENTRY_DIR = Path(__file__).resolve().parent
ENTRY_ID = ENTRY_DIR.name
ROOT = ENTRY_DIR.parents[1]
REFERENCE_DIR = ENTRY_DIR.parent / "go-0.12"

# The reference server's capabilities (go-0.12's manifest server.features); a
# scenario whose `requires` names anything else is reported as skipped, like
# the orchestrator would.
FEATURES = {"allowlist", "replay", "epoch", "bearer"}

# Outcome is the client's parsed one-line report (CONTRACT.md): HTTP status,
# §7 reject code (None on accept), the accepted identity (None on reject),
# and — when true — the chain-negotiation signal on the final response.
Outcome = dict[str, Any]


def fatal(msg: str) -> NoReturn:
    raise SystemExit(f"smoke: {msg}")


def build_reference_server(workdir: Path) -> Path:
    """Build the go-0.12 reference server the client is proven against."""
    binary = workdir / "server"
    proc = subprocess.run(
        ["go", "build", "-o", str(binary), "./cmd/server"],
        cwd=REFERENCE_DIR,
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        fatal(f"build go-0.12 server: {proc.stderr}")
    return binary


# --- driving this entry's client runnable ---

def run_client(sc: dict[str, Any], addr: str) -> list[Outcome]:
    """Invoke the entry's client once per repeat attempt with the scenario's
    contract flags and parse its outcome lines. The client must exit 0
    whether the request was accepted or rejected."""
    binary = Path(sys.executable).parent / f"{ENTRY_ID}-client"
    args = [
        str(binary), "--transport", "http", "--addr", addr,
        "--creds", str(ROOT / "fixture" / "creds" / sc["creds"]),
    ]
    if sc.get("nonce"):
        args += ["--nonce", sc["nonce"]]
    if sc["mode"] == "message":
        args += ["--mode", "message"]
        if sc.get("audience"):
            args += ["--audience", sc["audience"]]
        if sc.get("payload"):
            args += ["--payload", str(ROOT / sc["payload"])]
        if sc.get("tamper_payload"):
            args += ["--tamper-payload", str(ROOT / sc["tamper_payload"])]
        if sc.get("ttl"):
            args += ["--ttl=" + str(sc["ttl"])]
        if sc.get("chain"):
            args += ["--chain", sc["chain"]]

    outs: list[Outcome] = []
    for _ in range(sc.get("repeat") or 1):
        proc = subprocess.run(args, capture_output=True, text=True)
        if proc.returncode != 0:
            fatal(f"client {args}: exit {proc.returncode}: {proc.stderr}")
        try:
            outs.append(json.loads(proc.stdout))
        except json.JSONDecodeError as exc:
            fatal(f"client output {proc.stdout!r}: {exc}")
    return outs


# --- judging (mirrors the orchestrator's judge) ---

def judge(sc: dict[str, Any], outs: list[Outcome]) -> str | None:
    """Compare the outcomes to the scenario expectation; every pre-final
    repeat attempt must be accepted. Returns the mismatch, or None."""
    want = sc.get("expect_last") or sc["expect"]
    for out in outs[:-1]:
        if out["status"] != 200:
            return f"warmup expected accept, got {out}"
    out = outs[-1]
    got_chain = bool(out.get("chain_required"))
    if "chain_required" in want and got_chain != bool(want["chain_required"]):
        return f"chain_required = {got_chain}, want {want['chain_required']}"
    if want.get("accept"):
        if out["status"] != 200:
            return f"expected accept, got {out}"
        if out["reason"] is not None:
            return f"accept carried reason {out['reason']!r}"
        if out["identity"] is None:
            return "accept carried no identity"
        if "tenant" in want and out["identity"]["tenant"] != want["tenant"]:
            return f"tenant = {out['identity']['tenant']!r}, want {want['tenant']!r}"
        if "user" in want and out["identity"]["user"] != want["user"]:
            return f"user = {out['identity']['user']!r}, want {want['user']!r}"
        return None
    if out["status"] != 401:
        return f"expected reject with reason {want.get('reason')!r}, got {out}"
    if out["reason"] != want.get("reason"):
        return f"reason = {out['reason']!r}, want {want.get('reason')!r}"
    if out["identity"] is not None:
        return f"reject carried identity {out['identity']}"
    return None


# --- reference server lifecycle ---

@contextmanager
def server(binary: Path, mode: str) -> Iterator[str]:
    """Start the reference server, yield the bound address from its ready
    line, and assert the clean SIGTERM exit."""
    proc = subprocess.Popen(
        [
            str(binary), "--transport", "http", "--addr", "127.0.0.1:0",
            "--operator", str(ROOT / "fixture" / "operator.pub"),
            "--allowlist", str(ROOT / "fixture" / "allowlist.txt"),
            "--mode", mode,
        ],
        stdout=subprocess.PIPE,
        text=True,
    )
    try:
        assert proc.stdout is not None  # PIPE above
        line = _readline(proc.stdout, timeout=30.0)
        if not line.startswith("ready "):
            fatal(f"reference server ({mode}) did not report readiness: {line!r}")
        yield line.removeprefix("ready ").strip()
    finally:
        proc.send_signal(signal.SIGTERM)
        try:
            code = proc.wait(timeout=10)
        except subprocess.TimeoutExpired:
            proc.kill()
            fatal(f"reference server ({mode}) did not exit on SIGTERM")
        if code != 0:
            fatal(f"reference server ({mode}) exited {code} on SIGTERM, want 0")


def _readline(stream: Any, timeout: float) -> str:
    got: dict[str, str] = {}
    reader = threading.Thread(target=lambda: got.update(line=stream.readline()), daemon=True)
    reader.start()
    reader.join(timeout)
    return got.get("line", "")


def main() -> int:
    suite = yaml.safe_load((ROOT / "scenarios.yaml").read_text(encoding="utf-8"))
    applicable = [s for s in suite["scenarios"] if s.get("transport") in (None, "http")]
    passed = failed = skipped = 0
    with tempfile.TemporaryDirectory() as tmp:
        binary = build_reference_server(Path(tmp))
        for mode in ("signed", "message"):
            subset = [s for s in applicable if s.get("mode") == mode]
            if not subset:
                continue
            with server(binary, mode) as addr:
                for sc in subset:
                    if not set(sc.get("requires") or ()) <= FEATURES:
                        print(f"SKIP {sc['id']} (requires {sc['requires']})")
                        skipped += 1
                        continue
                    err = judge(sc, run_client(sc, addr))
                    if err is None:
                        print(f"PASS {sc['id']}")
                        passed += 1
                    else:
                        print(f"FAIL {sc['id']}: {err}")
                        failed += 1
    print(f"{ENTRY_ID} smoke: {passed} passed, {failed} failed, {skipped} skipped")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())

File: harness/go-0.12/smoke_test.go
// Package smoke proves the go-0.12 harness end-to-end: it builds both
// runnables, starts the server against the committed fixture, and drives the
// client through the scenarios.yaml suite on each transport and mode,
// asserting the contract outcomes (CONTRACT.md) including the §7 reason
// codes on rejection.
package smoke

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"
	"testing"
	"time"
)

// outcome mirrors the client's one-line report.
type outcome struct {
	Status   any              `json:"status"`
	Reason   *string          `json:"reason"`
	Identity *outcomeIdentity `json:"identity"`
}

type outcomeIdentity struct {
	Tenant string  `json:"tenant"`
	User   *string `json:"user"`
}

// bins builds the server and client once per test binary.
func bins(t *testing.T) (server, client string) {
	t.Helper()
	dir := t.TempDir()
	server = filepath.Join(dir, "server")
	client = filepath.Join(dir, "client")
	for target, pkg := range map[string]string{server: "./cmd/server", client: "./cmd/client"} {
		out, err := exec.Command("go", "build", "-o", target, pkg).CombinedOutput()
		if err != nil {
			t.Fatalf("build %s: %v\n%s", pkg, err, out)
		}
	}
	return server, client
}

func fixtureDir(t *testing.T) string {
	t.Helper()
	dir, err := filepath.Abs(filepath.Join("..", "..", "fixture"))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(filepath.Join(dir, "operator.pub")); err != nil {
		t.Fatalf("fixture not found (run fixture/gen): %v", err)
	}
	return dir
}

// startServer launches the harness server, waits for its readiness line, and
// registers a cleanup that SIGTERMs it and asserts a clean exit.
func startServer(t *testing.T, bin, transport, mode string) (addr string) {
	t.Helper()
	fixture := fixtureDir(t)
	cmd := exec.Command(bin,
		"--transport", transport,
		"--addr", "127.0.0.1:0",
		"--operator", filepath.Join(fixture, "operator.pub"),
		"--allowlist", filepath.Join(fixture, "allowlist.txt"),
		"--mode", mode,
	)
	cmd.Stderr = os.Stderr
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		t.Fatal(err)
	}
	if err := cmd.Start(); err != nil {
		t.Fatal(err)
	}

	ready := make(chan string, 1)
	go func() {
		scanner := bufio.NewScanner(stdout)
		if scanner.Scan() {
			ready <- scanner.Text()
		}
		close(ready)
	}()
	select {
	case line, ok := <-ready:
		if !ok || !strings.HasPrefix(line, "ready ") {
			t.Fatalf("server did not report readiness, got %q", line)
		}
		addr = strings.TrimPrefix(line, "ready ")
	case <-time.After(10 * time.Second):
		_ = cmd.Process.Kill()
		t.Fatal("server did not report readiness in time")
	}

	t.Cleanup(func() {
		if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
			t.Errorf("signal server: %v", err)
			return
		}
		if err := cmd.Wait(); err != nil {
			t.Errorf("server did not exit cleanly on SIGTERM: %v", err)
		}
	})
	return addr
}

// runClient invokes the harness client once and parses its outcome line. The
// client must exit 0 whether the request was accepted or rejected.
func runClient(t *testing.T, bin string, args ...string) outcome {
	t.Helper()
	cmd := exec.Command(bin, args...)
	cmd.Stderr = os.Stderr
	raw, err := cmd.Output()
	if err != nil {
		t.Fatalf("client %v: %v", args, err)
	}
	var out outcome
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("client output %q: %v", raw, err)
	}
	return out
}

// wantAccept asserts the accept outcome for a transport: HTTP 200 or gRPC
// OK, no reason, and the expected identity.
func wantAccept(t *testing.T, transport string, out outcome, tenant string, user *string) {
	t.Helper()
	wantStatus(t, transport, out, 200, "OK")
	if out.Reason != nil {
		t.Errorf("accept carried reason %q", *out.Reason)
	}
	if out.Identity == nil {
		t.Fatal("accept carried no identity")
	}
	if out.Identity.Tenant != tenant {
		t.Errorf("tenant = %q, want %q", out.Identity.Tenant, tenant)
	}
	switch {
	case user == nil && out.Identity.User != nil:
		t.Errorf("user = %q, want null", *out.Identity.User)
	case user != nil && (out.Identity.User == nil || *out.Identity.User != *user):
		t.Errorf("user = %v, want %q", out.Identity.User, *user)
	}
}

// wantReject asserts the reject outcome: HTTP 401 or gRPC UNAUTHENTICATED,
// the expected §7 reason, and no identity.
func wantReject(t *testing.T, transport string, out outcome, reason string) {
	t.Helper()
	wantStatus(t, transport, out, 401, "UNAUTHENTICATED")
	if out.Reason == nil || *out.Reason != reason {
		t.Errorf("reason = %v, want %q", out.Reason, reason)
	}
	if out.Identity != nil {
		t.Errorf("reject carried identity %+v", out.Identity)
	}
}

func wantStatus(t *testing.T, transport string, out outcome, httpWant int, grpcWant string) {
	t.Helper()
	switch transport {
	case "http":
		got, ok := out.Status.(float64)
		if !ok || int(got) != httpWant {
			t.Errorf("status = %v, want %d", out.Status, httpWant)
		}
	case "grpc":
		got, ok := out.Status.(string)
		if !ok || got != grpcWant {
			t.Errorf("status = %v, want %q", out.Status, grpcWant)
		}
	}
}

func strp(s string) *string { return &s }

func TestSigned(t *testing.T) {
	server, client := bins(t)
	fixture := fixtureDir(t)
	credsPath := func(name string) string { return filepath.Join(fixture, "creds", name) }

	for _, transport := range []string{"http", "grpc"} {
		t.Run(transport, func(t *testing.T) {
			addr := startServer(t, server, transport, "signed")
			call := func(creds string, extra ...string) outcome {
				args := append([]string{
					"--transport", transport, "--addr", addr,
					"--creds", credsPath(creds),
				}, extra...)
				return runClient(t, client, args...)
			}

			t.Run("account-valid", func(t *testing.T) {
				wantAccept(t, transport, call("account.creds"), "acme", nil)
			})
			t.Run("user-valid", func(t *testing.T) {
				wantAccept(t, transport, call("user.creds"), "acme", strp("alice"))
			})
			t.Run("bearer-valid", func(t *testing.T) {
				wantAccept(t, transport, call("bearer.creds"), "acme", strp("alice"))
			})
			t.Run("account-revoked", func(t *testing.T) {
				wantReject(t, transport, call("revoked.creds"), "not_allowlisted")
			})
			t.Run("expired", func(t *testing.T) {
				wantReject(t, transport, call("expired.creds"), "expired")
			})
			t.Run("wrong-key", func(t *testing.T) {
				wantReject(t, transport, call("wrongkey.creds"), "bad_request_signature")
			})
			t.Run("replay", func(t *testing.T) {
				nonce := fmt.Sprintf("interop-nonce-%s", transport)
				wantAccept(t, transport, call("user.creds", "--nonce", nonce), "acme", strp("alice"))
				wantReject(t, transport, call("user.creds", "--nonce", nonce), "replay")
			})
		})
	}
}

func TestMessage(t *testing.T) {
	server, client := bins(t)
	fixture := fixtureDir(t)

	for _, transport := range []string{"http", "grpc"} {
		t.Run(transport, func(t *testing.T) {
			addr := startServer(t, server, transport, "message")
			call := func(audience string) outcome {
				return runClient(t, client,
					"--transport", transport, "--addr", addr,
					"--creds", filepath.Join(fixture, "creds", "user.creds"),
					"--mode", "message",
					"--audience", audience,
					"--payload", filepath.Join(fixture, "payloads", "hello.bin"),
				)
			}

			t.Run("valid", func(t *testing.T) {
				wantAccept(t, transport, call("interop://sink"), "acme", strp("alice"))
			})
			t.Run("wrong-audience", func(t *testing.T) {
				wantReject(t, transport, call("interop://other"), "wrong_audience")
			})
		})
	}
}
